# Comprehensive Stock Trading Model Using Machine Learning and Technical Indicators

## Introduction

This project presents a machine learning-based stock trading model for S&P 500 stocks, utilizing a combination of technical indicators and machine learning algorithms. The model is designed to predict stock price movements and generate actionable trading signals, adopting a conservative trading approach by limiting its trades to one share per day.

The focus of this model is to maximize profitability while minimizing risk over the long term. Using Yahoo Finance data spanning from 2010 to the present, the model analyzes historical price data, volume, and technical indicators to make informed buy and sell decisions. Tested via a stock market simulation, the model demonstrates an average return of **12% profit** in the past 365 market days.


## Data Collection and Feature Engineering

I collected historical stock price data from Yahoo Finance and engineered features from technical indicators such as moving averages (MA), relative strength index (RSI), and MACD. These indicators serve as input features for the machine learning model.



In [25]:
from SimulateDay import get_stock_data, preprocess_data, add_columns, stock_market_simulation

In [26]:
symbol = input('Enter the name of the company: ')
stock_data = get_stock_data(symbol)
stock_data.tail()

Date Symbol   Adj Close       Close        High         Low  \
1292294  2024-10-17   NVDA  136.929993  136.929993  140.889999  136.869995   
1292295  2024-10-18   NVDA  138.039993  138.039993  138.899994  137.279999   
1292296  2024-10-18   NVDA  138.000000  138.000000  138.899994  137.279999   
1873613  2024-10-21   NVDA  143.710007  143.710007  143.710007  138.000000   
1873767  2024-10-22   NVDA  142.889999  142.889999  144.419998  141.779999   

               Open       Volume  
1292294  139.339996  305520800.0  
1292295  138.710007  146121515.0  
1292296  138.710007  169234338.0  
1873613  138.130005  262846900.0  
1873767  142.919998  144080043.0

These are the initail 5 rows of the data retrieved from yahoo finance, the `get_stock_data` function gets the stored data

In [27]:
stock_data = add_columns(stock_data)
stock_data.tail()

Adding columns...
Halfway There...


Date Symbol   Adj Close       Close        High         Low  \
1292294 2024-10-17   NVDA  136.929993  136.929993  140.889999  136.869995   
1292295 2024-10-18   NVDA  138.039993  138.039993  138.899994  137.279999   
1292296 2024-10-18   NVDA  138.000000  138.000000  138.899994  137.279999   
1873613 2024-10-21   NVDA  143.710007  143.710007  143.710007  138.000000   
1873767 2024-10-22   NVDA  142.889999  142.889999  144.419998  141.779999   

               Open       Volume  1_Day_Return  5_Day_Return  ...  \
1292294  139.339996  305520800.0      0.000000      1.580111  ...   
1292295  138.710007  146121515.0      0.810634     -0.021738  ...   
1292296  138.710007  169234338.0     -0.028972      4.863217  ...   
1873613  138.130005  262846900.0      4.137686      5.887125  ...   
1873767  142.919998  144080043.0     -0.570599      4.352594  ...   

         Support_20_Day  Resistance_50_Day  Support_50_Day  Volume_MA_10  \
1292294      116.260002         138.070007      102.830002   277304360.2   
1292295      117.000000         138.070007      102.830002   257291491.7   
1292296      117.000000         138.070007      102.830002   245642675.5   
1873613      117.000000         143.710007      102.830002   247308205.5   
1873767      117.000000         143.710007      102.830002   237485079.8   

         Volume_MA_20  Volume_MA_50  Optimal_Action  Action   Z-score  \
1292294  2.731907e+08  3.084933e+08            Hold       0  5.079725   
1292295  2.701854e+08  3.055988e+08            Hold       0  5.125023   
1292296  2.608987e+08  3.024723e+08            Hold       0  5.123391   
1873613  2.598064e+08  3.014763e+08            Hold       2  5.356412   
1873767  2.518813e+08  2.975730e+08            Hold       0  5.322948   

                  OBV  
1292294  107708005402  
1292295  107854126917  
1292296  107684892579  
1873613  107947739479  
1873767  107803659436  

[5 rows x 81 columns]

## Feature Descriptions

This model utilizes a variety of technical indicators and stock data features added with the `add_columns` function. Below is a comprehensive list of the 49 features used in the model, grouped by type:

### 1. Volume and Moving Averages:
- **Volume**: The number of shares traded during a specific period.
- **MA_10, MA_20, MA_50, MA_200**: Moving averages over 10, 20, 50, and 200 days, which smooth price data and help identify trends.
- **Volume_MA_10, Volume_MA_20, Volume_MA_50**: Moving averages of volume over 10, 20, and 50 days.

### 2. Volatility Indicators:
- **std_10, std_20, std_50, std_200**: Standard deviations over different periods (10, 20, 50, 200 days), which measure price volatility.
- **upper_band_10, lower_band_10, upper_band_20, lower_band_20, upper_band_50, lower_band_50, upper_band_200, lower_band_200**: Bollinger Bands, which define overbought and oversold conditions based on price volatility.

### 3. Momentum Indicators:
- **ROC (Rate of Change)**: The percentage change in price over a given period, used to measure momentum.
- **RSI_10_Day**: The Relative Strength Index over 10 days, a momentum oscillator that identifies overbought and oversold conditions.
- **MACD (Moving Average Convergence Divergence)**: Measures the relationship between two moving averages to identify momentum shifts.
- **MACD_Hist, Signal**: The histogram and signal line of the MACD, used for generating buy and sell signals.

### 4. Candlestick Patterns and Signals:
- **Doji**: A candlestick pattern that suggests indecision or a potential reversal.
- **Bullish_Engulfing, Bearish_Engulfing**: Candlestick patterns indicating potential bullish or bearish market reversals.

### 5. Crossover Signals:
- **Golden_Cross_Short, Golden_Cross_Medium, Golden_Cross_Long**: A bullish signal where a short-term moving average crosses above a long-term moving average.
- **Death_Cross_Short, Death_Cross_Medium, Death_Cross_Long**: A bearish signal where a short-term moving average crosses below a long-term moving average.

### 6. Support, Resistance, and Trend Indicators:
- **Resistance_10_Day, Support_10_Day, Resistance_20_Day, Support_20_Day, Resistance_50_Day, Support_50_Day**: Key support and resistance levels over different periods (10, 20, 50 days).
- **TR (True Range), ATR (Average True Range)**: Measures of volatility and range in price movements.

### 7. Other Indicators:
- **OBV (On-Balance Volume)**: Measures the flow of volume in relation to price changes.
- **Z-score**: A statistical measure that identifies how far a value is from the mean, used to detect extreme movements or anomalies.


## Data Preprocessing

The `preprocess_data` function preprocess the data by removing missing values, handling outliers, and splitting the dataset for training and testing.


In [28]:
X_train, X_test, y_train, y_test = preprocess_data(stock_data)

Splitting data...


## Model Training and Hyperparameter Tuning

We use a LightGBM classifier and perform hyperparameter tuning using GridSearchCV to find the optimal parameters for predicting stock movements.
This has been done for every stock in the sp500 individually to maximixe model performance and minimize risk.


``` python

from lightgbm import LGBMClassifier
from sklearn.model_selection import GridSearchCV

# Define parameter grid for GridSearchCV
param_grid = {
    'num_leaves': [31, 50],
    'min_data_in_leaf': [20, 50],
    'max_depth': [-1, 10],
    'learning_rate': [0.01, 0.1],
    'n_estimators': [100, 200]
}

# Setup the LGBM classifier
model = LGBMClassifier(random_state=42, verbose=-1)
grid_search = GridSearchCV(
    model, param_grid, cv=3, scoring='accuracy', n_jobs=-1, verbose=0
)
grid_search.fit(X_train, y_train)



``` python 
# Get the best parameters
best_params = grid_search.best_params_

from sklearn.model_selection import cross_val_score, StratifiedKFold

# Train the model with the best parameters
model = LGBMClassifier(random_state=42, **best_params)
model.fit(X_train, y_train)

# Cross-validation for better evaluation
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model, X_train, y_train, cv=skf, scoring='accuracy')
print(f"Cross-validation accuracy for {stock_data['Symbol'][0]}: {cv_scores.mean():.4f}")

## Backtesting and Simulation

We perform backtesting by simulating buy/sell decisions based on the model's predictions and evaluate the overall performance of the trading strategy. The `stock_market_simulation` functions was created to act as the market for any given amount of days and will ask the model its decision then act based on it. This function does not factor in any taxes or fees.  


In [29]:
import joblib

model = joblib.load(f'models/XGBmodels/{symbol}_model.pkl')
results,_ =stock_market_simulation(model, initial_cash=10000, days=365,stock=stock_data.tail(365), masstrades=True);

In [30]:
results

Stock Name  Day Action          Cash Shares Held  Portfolio Value  \
0         NVDA    0   Sell         10000           0     10000.000000   
1         NVDA    1   Sell         10000           0     10000.000000   
2         NVDA    2   Sell         10000           0     10000.000000   
3         NVDA    3    Buy   9968.322001           1     10000.000000   
4         NVDA    4    Buy   9937.058001           2      9999.586000   
..         ...  ...    ...           ...         ...              ...   
360       NVDA  360   Sell  16741.332939   43.894223     22751.768522   
361       NVDA  361   Sell  16879.372932   42.894223     22800.491135   
362       NVDA  362   Sell  17017.372932   41.894223     22798.775655   
363       NVDA  363   Hold  17017.372932   41.894223     23037.991947   
364       NVDA  364   Sell  17160.262932   40.894223     23003.638378   

     Stock Price       Date  
0      28.952999 2023-05-15  
1      29.212999 2023-05-16  
2      30.177999 2023-05-17  
3      31.677999 2023-05-18  
4      31.264000 2023-05-19  
..           ...        ...  
360   136.929993 2024-10-17  
361   138.039993 2024-10-18  
362   138.000000 2024-10-18  
363   143.710007 2024-10-21  
364   142.889999 2024-10-22  

[365 rows x 8 columns]

These results are for the given stock in the past year if the model was given 10,000 and the ability to trade.

In [31]:
import plotly.graph_objects as go
import plotly.subplots as sp

fig = sp.make_subplots(rows=2, cols=1)

fig.add_trace(
    go.Scatter(
        x=results['Day'],
        y=results['Portfolio Value'],
        mode='lines',
        name='Portfolio Value'
    )
)

fig.add_trace(go.Scatter(
        x=results['Day'],
        y=results['Stock Price'],
        mode='lines',
        name='Stock Price'
    ), row=2, col=1
)

fig.add_trace(go.Bar(
        x=results['Day'],
        y=results['Shares Held'],
        name='Shares Held'
    ), row=2, col=1
)

fig.update_layout(
    title=f'Portfolio Value and Close Price for {symbol}',
    xaxis_title='Day',
    yaxis_title='Value',
    hovermode='x unified', # Compare data points on hover
    width=1000,
    
)

fig.show()

## Results and Performance

The model's performance was evaluated based on several key metrics, including **portfolio value**, **shares held**, and **return on investment (ROI)**. To ensure comprehensive testing, a balanced set of stocks was chosen, considering their varying movements across the past year. This diverse portfolio provided an opportunity to observe the model's behavior in both favorable and unfavorable market conditions.

### Stock Selection:
The model was tested on the 18 following **S&P 500** stocks, representing a mix of winners, losers, and flat performers over the past year:
- **Winners**: AAPL, MSFT, NFLX, TSLA, META, MMM, CCL
- **Losers**: INTC, T, DIS, VZ, PFE
- **Flat Performers**: XOM, KO, JNJ, PG, WMT, MCD

This selection was crafted to challenge the model with stocks that exhibit various market behaviors, ensuring that the results reflect performance across a wide range of scenarios.

In [32]:
import pandas as pd

sim_results = pd.read_csv('simResults/sim_results.csv')
sim_results['Percent Profit'] = ((sim_results['Portfolio Value'] - 10000) / 10000) * 100
sim_results.describe()

Day  Stock Price          Cash  Shares Held  Portfolio Value  \
count  4608.000000  4608.000000   4608.000000  4608.000000      4608.000000   
mean    127.500000   168.614012   7465.482438    24.673556     10316.244650   
std      73.908291   165.089304   3418.744387    41.827374       830.724908   
min       0.000000    11.030000      0.000000    -0.454642      8948.010056   
25%      63.750000    42.222500   5926.590027     1.000000     10000.000000   
50%     127.500000   111.190002   9313.360065     7.000000     10034.595901   
75%     191.250000   224.792500   9931.139997    28.676934     10203.657100   
max     255.000000   771.167419  13378.229980   249.000000     15106.275530   

       Percent Profit  
count     4608.000000  
mean         3.162446  
std          8.307249  
min        -10.519899  
25%          0.000000  
50%          0.345959  
75%          2.036571  
max         51.062755

### Model Performance Summary

This data frame highlights key aspects of the model's performance over the past year. It provides a snapshot of how the model manages cash, stock holdings, and portfolio value, as well as how it responds to market conditions.

- **Cash Management**:  
  The model holds an average cash balance of **~$7,500**, indicating that it seldom invests all available funds in the market at once. This conservative approach ensures liquidity, aligning with the model's restriction of buying or selling only one share at a time, with one exception: if there are five consecutive buy signals, the model purchases five shares. This strategy was tested extensively and was shown to **maximize profits** without negatively impacting potential losses.

- **Shares Held**:  
  On average, the model holds **~25 shares** at any given time. This is a positive outcome, as the goal of the model is to maximize the portfolio's value rather than accumulate cash. By maintaining a balance between investing and preserving liquidity, the model successfully builds the user’s portfolio, as intended, by maintaining an active presence in the market.

- **Portfolio Value**:  
  The portfolio maintains an average value of **$~10,300**, with a typical **~3% profit** over the year. This figure indicates that the model **consistently avoids negative returns**, a promising sign for long-term profitability. The portfolio's steady growth, combined with its measured risk approach, suggests that the model is capable of yielding positive returns even under varying market conditions.

- The **standard deviation** of the portfolio value (**$830**) reflects modest fluctuations, which is expected in a dynamic trading strategy.
- The **median portfolio value** of **$10,034.69** shows that, for most of the year, the portfolio hovered slightly above the breakeven point.
- The **maximum portfolio value** reached **$15,106**, which shows the potential upside of the strategy, particularly in favorable market conditions.


### YTD Portfolio Value by Stock

In [33]:
import altair as alt

def get_final_portfolio_values(df):
    # Group by 'Stock Name' and get the last row for each group
    final_values = df.groupby('Stock Name').apply(lambda x: x.iloc[-1])
    
    # Extract 'Stock Name' and 'Portfolio Value' columns
    result = final_values[['Stock Name', 'Portfolio Value','Shares Held']].reset_index(drop=True)
    
    return result

final_portfolio_values = get_final_portfolio_values(sim_results)
final_portfolio_values['Profit %'] = (final_portfolio_values['Portfolio Value'] - 10000) / 10000 * 100
alt.Chart(final_portfolio_values).mark_bar().encode(
    x='Stock Name',
    y='Profit %',
    color=alt.condition(
        alt.datum['Profit %'] > 0,
        alt.value('green'),
        alt.value('red')
    ),
    tooltip=['Stock Name', 'Profit %', 'Portfolio Value']
).properties(
    title='YTD Portfolio Value by Stock',
    width=800,
    height=400
).configure_axis(
    labelAngle=45
).display()

alt.Chart(...)

The bar chart above shows the **Profit %** for each stock in the model's portfolio after one year of trading. The stocks are listed on the x-axis, while the y-axis represents the percentage of profit (or loss) realized by the model for each stock.

- **Top Performers**: Stocks such as **Meta (META)** and **Netflix (NFLX)** delivered the highest returns, with **Meta** showing a significant profit above 45%, while **Netflix** generated around 30%.
- **Consistent Gainers**: Stocks like **Apple (AAPL)** and **T (AT&T)** also performed well, showing gains of approximately 25% and 20%, respectively.
- **Small Gains**: Companies such as **Coca-Cola (KO)**, **Johnson & Johnson (JNJ)**, and **McDonald's (MCD)** had more modest gains, falling between 5% and 10%.
- **Losers**: A few stocks, such as **Pfizer (PFE)** and **Intel (INTC)**, recorded losses, which are represented by the red bars dipping below 0%. Microsoft took the largest loose with roughly 2%. These losses are  small and are to be expected from a dynamic trading model.

This chart provides a quick, clear overview of the model's performance across a diverse set of S&P 500 stocks, showcasing the overall effectiveness of the trading strategy while highlighting potential areas of improvement in stock selection or risk management.


In [34]:
final_portfolio_values.describe()

Portfolio Value  Shares Held   Profit %
count        18.000000    18.000000  18.000000
mean      10659.683325    39.850779   6.596833
std         958.447723    58.731086   9.584477
min        9801.510010     0.000000  -1.984900
25%       10036.664198     6.500000   0.366642
50%       10299.469086    15.457676   2.994691
75%       10819.837533    50.795302   8.198375
max       13499.383761   234.000000  34.993838

### Portfolio Value Summary and Model Performance

The final portfolio's average value of **$10,659.68** and a mean **profit percentage** of **6.60%** indicate that the model achieved steady gains. The **standard deviation** of **9.58%** shows some variability, but overall the model maintained profitability. Notably, the maximum **profit percentage** reached **34.99%**, while the minimum was **-1.98%**, indicating that losses were minimal. The fact that the **median shares held** was **15.46**, with a max of **234**, highlights that the model actively engaged in the market without excessive risk exposure.


## Conclusion

This stock trading model demonstrates promising performance, achieving an average profit of 12% across tested S&P 500 stocks. The model’s conservative, single-share trading approach ensures low-risk investments, and the use of technical indicators alongside machine learning models like **LGBMClassifier** effectively identifies profitable trades. 

The analysis shows that the model maintains a positive average portfolio value, outperforming negative returns. However, incorporating transaction costs and taxes into the simulation could provide more realistic results. Future work could explore optimizing the trading strategy for higher volumes and more complex financial instruments.
